README.md
# Accessible Accordions & disclosure widgets   
At its essence, an accordion component consists of a series of related [disclosure widgets](http://w3c.github.io/aria-practices/#disclosure) (aka toggle buttons that show/hide their related content). These accordion widgets are visually related to each other, and are grouped siblings in the DOM (as either a series of content sections within a wrapper element, or within a list element). Often accordions are programmatically aware of their siblings' current state, though this is optional behavior.

* [Accordion Demo Page](https://scottaohara.github.io/a11y_accordions/)  
* Based on [ARIA Authoring Practices](https://w3c.github.io/aria-practices/#accordion)  
* [Accessible Accordions blog article](https://www.scottohara.me/blog/2017/10/25/accordion-release.html)  

## Install
You can get this package [on npm](https://www.npmjs.com/package/a11y_accordions):
```
$ npm i a11y_accordions
```

Clone it:
```
$ git clone https://github.com/scottaohara/a11y_accordions.git
```

Or [download a zip of the repository](https://github.com/scottaohara/a11y_accordions/archive/master.zip).

The CSS for this component is included in `assets/css/`. The classes are added to the accordion markup when the script/page loads. Presently, changing the class names requires a change in the JavaScript file and the CSS. Adjust the styles as necessary for your project.


## Minimum Required Mark-up  
```html
<div data-aria-accordion>
  <h3 data-aria-accordion-heading>
    Heading Here
  </h3>
  <div data-aria-accordion-panel>
    <p>
      Content here
    </p>
  </div>
  <!-- 
    repeat for as many accordion panels needed
  -->
</div>
```

### Under the hood
The `data-aria-accordion` attribute is the key for initiating the process to convert the minimum markup into a functioning accordion component.  

If an `id` is not pre-set on the accordion wrapper, then one will be auto generated. An ID is necessary to serve as the basis for populating generated IDs onto the child accordion panels. More on why these IDs are needed, later.

When an accordion is identified, the setup script continues to run and identify each heading and panel within the accordion. The different `data-aria-accordion-` attributes are necessary for the setup process to run and appropriately identify the key elements of the accordion.

#### During the setup process the following occurs:  
* The classes for an accordion, and its child headings and panels will be added, corresponding to the appropriate `data-` attributes.
* The panels are hidden, and if a default panel was set (see options) an attribute of `aria-hidden="false"` will be set to that panel.  
* The `id` of the accordion container is used as the basis to generate unique IDs for each of the panels of the accordion.  
* a `<button>` is dynamically created and inserted into the `data-aria-accordion-heading` element. The previous text of the heading is then inserted into this button. Presently, this strips out any previous HTML tags that were in the heading and only retains the text string.  
* The `<button>` is given the class `accordion__trigger`, an `aria-controls` attribute, with the value set to the panel `id` that appears next in the DOM. While `aria-controls` is not presently useful in all assistive technologies, JAWS users will be given additional keys to press to begin interacting with the associated content.   
* Depending on the initial state of the associated panel, e.g. is it set to be opened by default or not, the `<button>` will also receive the attribute `aria-expanded`, which will either have the value `true` or `false`. These value will indicate the current state of their associated panels, to uses of assistive technologies.   
* If an accordion has the `data-constant` attribute set (see options), then the `<button>` with `aria-expanded="true"` will also have an `aria-disabled="true"` set to it.  


## Options  
The following `data` attributes can be added to an accordion instance to alter the default accordion setup settings. None of the `data` attributes affect nested accordions, so a parent accordion that requires a consistently open panel would not pass down that requirement to a child accordion.  

### `data-default` 
* Leave off or set to "none" to not have a panel open by default.  
* Set the attribute, with no value, to open the first panel by default. Set the value to the panel number ("1" for the first panel in the DOM, 2 for the second, etc.) to open that specific panel by default.  
* Setting a negative number, or 0 will open the first panel by default.  
* Setting a number greater than the number of panels of an accordion, will open the last panel by default.  

### `data-constant`  
By placing this attribute on the accordion wrapper, it will indicate that an accordion panel must always be expanded. An `aria-disabled="true"` will be set to the `button` that controls the expanded panel. This attribute will indicate to assistive technologies that the `button` can not be currently interacted with.    

### `data-transition`  
Using this attribute on an accordion will add a modifier class to each child accordion panel where CSS transitions can be applied to transition the panel from hidden to revealed state.

### `data-multi`  
This attribute requires no value. Setting it will allow multiple accordion panels, of a particular group, to be opened at once.  


## Keyboard Controls  
When focus on the accordion triggers:  
* <kbd>Tab</kbd> and <kbd>Shift</kbd> + <kbd>Tab</kbd>, or standard AT navigation keys are used to navigate between accordion triggers and any focusable content within an accordion panel.  
* <kbd>Home</kbd> moves focus to the first accordion trigger in the group.  
* <kbd>End</kbd> moves focus to the last accordion trigger in the group.  
* <kbd>Space</kbd> or <kbd>Enter</kbd> open/close the trigger's associated panel.  

### Note about VoiceOver
When navigating a document by headings, or list items (if an accordion is in a list), VoiceOver will announce the text of the element, the fact it contains a button, and the current state of the button (expanded or collapsed). It will then announce the heading level.

VoiceOver has its own special command to activating actionable elements: <kbd>Ctrl</kbd> + <kbd>Option</kbd> + <kbd>Space</kbd>, where screen readers like JAWS and NVDA do not.  If navigating by heading, the <kbd>Enter</kbd> or <kbd>Space</kbd> keys alone will not toggle the state of the accordion trigger. However the <kbd>Ctrl</kbd> + <kbd>Option</kbd> + <kbd>Space</kbd> keys will activate the trigger.  This is not a "bug", but just how VoiceOver works.  

If navigating by button element, or using <kbd>tab</kbd> to navigate the document, <kbd>Enter</kbd> and <kbd>Space</kbd> will activate the trigger on their own.


## License & Such  
This script was written by [Scott O'Hara](https://twitter.com/scottohara).

It has an [MIT license](https://github.com/scottaohara/accessible-components/blob/master/LICENSE.md).

Do with it what you will :)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
- Optional feature to add support for up/down arrow keys.
- Allow for custom component classes to be declared via a data-class attribute.

## [3.2.1] - 2018-06-28
### Changed
- Fixed instructions for CSS selectors in readme file.


## [3.2.0] - 2018-06-28
### Added
- package.json and registered as NPM package.
- CHANGELOG.md

### Changed
- `data-transition` attribute no longer allows for a value. Instead it merely adds a transition class to each accordion panel to allow for transitions for specific CSS properties to be customized in the CSS file.
- Instead of requiring `.js` prior to CSS selectors, remove this `.js` selector dependency and have the base markup utilize data attributes. When the script runs on page load, the classes that were there by default will now be applied by the script.

## [3.1.1] - 2018-06-11
### Added
- Add "arrows" for accordion trigger styling.


## [3.1.0] - 2018-02-09
### Changed
- [Modify the manner in which IDs are generated](https://github.com/scottaohara/a11y_accordions/commit/5723fafddac2dcbf102a0f99bd9f6d3b2e676dd1).


## [3.0.0] - 2018-02-09
### Added
- CSS to ensure list semantics are respected with Safari + VoiceOver.

### Changed
- Updated documentation and notes about screen reader usability.
- Allow accordions to have an `ol` or `ul` base markup pattern.
- Fix bug with multi-accordion panels.


## [2.0.1] - 2017-10-27
### Changed
- NaN bug fix.


## [2.0.0] - 2017-10-24
### Changed
- Converted script from jQuery to ES5 Vanilla JavaScript.

### Removed
- jQuery dependency
- CSS :target selector for no-js functionality. Instead the no-js functionality should just be the static underlying markup.


## 1.0.2 - 2017-10-24
### Added
- CSS :target selector for no-js functionality

### Changed
- Fix spacebar bug with Firefox
- Modified styling selectors
- Smarter selectors for nesting of accordion widgets
- Timeout function to mitigate Firefox animation issue


assets/a11y-accordion.js
;(function ( $, w, doc ) {

  'use strict';

  var a11yAccordion = {};

  a11yAccordion.NS = "a11yAccordion";
  a11yAccordion.AUTHOR = "Scott O'Hara";
  a11yAccordion.VERION = "1.0.2";
  a11yAccordion.LICENSE = "https://github.com/scottaohara/accessible-components/blob/master/LICENSE.md";

  // define the plug-in
  var accWidget  = '[data-action="is-accordion"]';

  $.fn.extend({

    a11yAccordion: function ( e ) {

      /*
        OPTIONS:

        add the following data attributes to their respected element
        to extend the accordion setup process

          container
            * data-showall="true"
                show all panels by default

            * data-multi-open="true"
                individual accordion can have multiple tabs
                open at once

          panel
            * data-showbydefault="true"
                show individual panel by default

            * data-tab-label="Your Label Here"
                set the text for the auto generated
                tab/trigger for this panel (only if the panel does not have
                a preceding heading element)

      */

      // setup global class variables
      var accTrigger    = '.accordion__trigger',
          accPanel      = '.accordion__panel';


      return this.each( function () {

        // set up variables specific to the each instance
        var id = this.id,
            $self = $('#' + id),
            // these two variables will be reused in various functions
            setFalse = { 'aria-selected': 'false', 'tabindex': '-1' },
            setTrue = { 'aria-selected': 'true', 'tabindex': '0' },


        genAcc = function () {

          // check to see if there are any accTriggers,
          // if not, they need to be generated
          if ( !$self.find('> ' + accTrigger).length ) {

            // start a counter
            var $panelNum = 1;

            // find the panels within this accordion instance
            $self.find('> ' + accPanel).each( function () {
              var $this = $(this);

              // now create an ID based on the instance ID + _panel_ + a randomly generated #
              $this.attr('id', id + '_panel_' + $panelNum );

              // now grab that ID for later
              var $grabID = $this.attr('id'),

                // The expected mark-up for a no-js accordion should have a heading
                // before each tab panel. So if that exists, then grab the text
                // from that heading to use as the tab trigger label
                // OR if a heading isn't there, check to see if a data-tab-label was
                // set to the tab-panel.
                // IF both of those checks fail, then just call it "tab" + it's
                // # designation in the tab panel set
                $grabLabel = $this.prev(':header').text() || $this.attr('data-tab-label') || 'Tab ' + $panelNum,

                // Put it all together as a new <li>tab</li>
                $createTabItem = '<button type="button" data-href="#'+$grabID+'" class="accordion__trigger">'+$grabLabel+'</button>';

                $this.before($createTabItem);

              return $panelNum = $panelNum + 1;

            });

          }

        }, // end genAcc


        // setup the accordion and it's tabs and panels
        setupAcc = function () {

          // setup the component to have a role of tablist
          $self.attr({ 'role': 'tablist' });

          // if there was a heading set for the no-js solution (there should have been)
          // then it can be removed because we've already grabbed the text from it
          // for use in the auto generated <a>
          $self.find('> :header').remove();

          // if an accordion can have multiple panels opened at a time,
          // then add this aria attribute
          if ( $self.attr('data-multi-open') ) {

            $self.attr('aria-multiselectable', 'true');

          }


          // find the panel triggers within each instance
          // generate the appropriate attributes and values
          $self.find('> ' + accTrigger).each( function () {

            var $this = $(this),
                $getURL = $this.attr('href') || $this.attr('data-href'),
                $thisTarget = $getURL.split('#')[1];

            $this.attr({
              'aria-controls': $thisTarget,
              'aria-expanded': 'false',
              'aria-selected': 'false',
              'id': $thisTarget + '_tab',
              'role': 'tab',
              'tabindex': '-1'
            });

          }); // end $self.find(accTrigger).each


          // setup the panels
          $self.find('> ' + accPanel).each( function () {

            var $this = $(this);

            // setup panel to have appropriate attribute hooks
            $this.attr({
              'aria-hidden': 'true',
              'aria-labelledby': $this.before().attr('id') + '_tab',
              'role': 'tabpanel'
            });

            // if a panel should be shown by default,
            // update the aria-expanded attribute and .show() it.
            // then update the paired trigger's attributes to
            // show it as active/selected
            if ( $this.attr('data-showbydefault') ) {

              // set the appropriate aria attribute values
              $this.attr({ 'aria-hidden': 'false' }).show(); // show this by default

              // and set the corresponding trigger to have
              // the correct attribute values as well
              $this.prev(accTrigger).attr({
                'aria-expanded': 'true',
                'aria-selected': 'true',
                'tabindex': '0'
              }).addClass('is-active');

              // and finally, add a class to the parent
              // to mark it as an accordion that has a
              // default tab
              $this.parent().addClass('has-default');
            }

          }); // end $self.find(accPanel).each


          // if a default tab wasn't set, then the first tab
          // needs to be focusable in the tablist
          if ( !$self.hasClass('has-default') ) {

            $self.find('> ' + accTrigger + ':first-of-type').attr({ 'tabindex': '0' });

          }

        }, // end setupAcc



        // show all panel resetup
        showAllSetup = function () {

          if ( $self.attr('data-showall') === 'true' ) {

            $self.find('> ' + accTrigger).attr('aria-expanded', 'true');
            $self.find('> ' + accPanel).attr('aria-hidden', 'false').show();
            $self.find('> ' + accTrigger).first().attr('aria-selected', 'true').addClass('is-active');

          }

        },




        // on panel click
        panelClick = function ( e ) {

          e.stopPropagation();

          var $e = $(e.target).closest(accPanel),
              $getTarget = $e.attr('aria-labelledby'),
              $target = $('#'+$getTarget);


          $self.find('> ' + accTrigger).attr(setFalse).removeClass('is-active');


          // run the check expanded function
          checkExpanded();


          $target.attr(setTrue).addClass('is-active').removeClass('was-active');

        },


        // check to see if a panel was previously expanded
        // as that means it used to be active, and we'll
        // need this class for later
        checkExpanded = function () {

          $self.find('> ' + accTrigger).each( function () {

            var $this = $(this);

            if ( $this.attr('aria-expanded') === 'true' ) {

              $this.addClass('was-active');

            }

          });

        },


        // trigger logic
        panelReveal = function ( e ) {

          // prevent browser jumping and the URI being
          // added to the browser's address bar
          e.preventDefault();

          var $e = $(e.target),
              $getTarget = $e.attr('aria-controls'),
              $target = $('#'+$getTarget);


          // run a check to see if the multiOpen flag is up
          // if not, then when a new tab is opened, previously
          // opened tags need to close
          if ( !$self.attr('data-multi-open') ) {

            $self.find('> ' + accTrigger).attr({
              'aria-selected': 'false',
              'aria-expanded': 'false',
              'tabindex': '-1'
            });

            $self.find('> ' + accPanel).attr({ 'aria-hidden': 'true' }).slideUp();

          }


          // run the check expanded function
          checkExpanded();


          // if a tab was previously active, if it is clicked again
          // it needs to close, but still become the currently active
          // and tabindexable element
          if ( $e.hasClass('was-active') ) {

            $self.find('> ' + accTrigger).removeClass('is-active').attr(setFalse);

            $e.attr({
              'aria-expanded': 'false',
              'aria-selected': 'true',
              'tabindex': '0'
            }).removeClass('was-active').focus();

            $target.attr('aria-hidden', 'true').slideUp();

            return;

          }


          // regardless of if multiOpen is true or false,
          // if the target trigger has a class of is-active,
          // and a user activates it again, then it should
          // close that panel and deactivate the tab
          if ( $e.hasClass('is-active') ) {

            $e.attr({
              'aria-expanded': 'false',
              'tabindex': '0'
            }).removeClass('is-active');

            $target.attr({ 'aria-hidden': 'true' }).slideUp();

            return;

          }


          else if ( !$e.hasClass('is-active') ) {

            $self.find('> ' + accTrigger).removeClass('is-active').attr(setFalse);

            $e.attr({
              'tabindex': '0',
              'aria-selected': 'true',
              'aria-expanded': 'true'
            }).addClass('is-active').focus();

            $target.attr({ 'aria-hidden': 'false' }).slideDown();

            return;

          }

        },



        // tab/trigger keyboard controls
        keytrolls = function ( e ) {

          var keyCode = e.which,
              $grabThisTrigger,
              $currentTabItem;

          // are we in the tab/headings or inside a panel?
          if ( $(e.target).attr('role') === 'tab' ) {
            $currentTabItem = $(e.target);
          }
          else if ( $(e.target).closest(accPanel) ) {
            $grabThisTrigger = $(e.target).closest(accPanel).attr('aria-labelledby'),
            $currentTabItem = $('#'+$grabThisTrigger);
          }


          // now that we know what our currentTabItem/Accordion are
          // we can set up some more variables
          var $currentAccordion = $currentTabItem.parent(),
              $firstTab = $currentAccordion.find('> ' + accTrigger).first(),
              $lastTab = $currentAccordion.find('> ' + accTrigger).last(),
              $prevTab = $currentTabItem.prev().prev(),
              $nextTab = $currentTabItem.next().next()


          if ( $currentTabItem.is(accTrigger + ':first-of-type') ) {
            $prevTab = $lastTab;
          }
          else if ( $currentTabItem.is(accTrigger + ':last-of-type') ) {
            $nextTab = $firstTab;
          }


          // now there are different keyboard controls for our different situations
          if ( $(e.target).attr('role') === 'tab' ) {

            switch ( keyCode ) {

              // right + down
              case 39: // right
              case 40: // down
                e.preventDefault();
                $currentTabItem.attr(setFalse);
                $nextTab.focus().attr(setTrue);
                break;

              // left + up
              case 37: // left
              case 38: // up
                e.preventDefault();
                $currentTabItem.attr(setFalse);
                $prevTab.focus().attr(setTrue);
                break;

              case 35: // end
                e.stopPropagation();
                $lastTab.focus().attr(setTrue);
                break;

              case 36: // home
                e.stopPropagation();
                $firstTab.focus().attr(setTrue);
                break;

              default:
                break;

            } // end switch


          if ( $(e.target).is('a') ) {
            switch ( keyCode ) {
              case 13: // enter
              case 32: // space bar
                e.stopPropagation();
                panelReveal( e );
                break;
            }
          }
          else if ( $(e.target).is('button') ) {
            switch ( keyCode ) {
              case 13: // enter
                e.stopPropagation();
                panelReveal( e );
                break;
            }
          }


          } // end if

          else if ( $(e.target).closest(accPanel) ) {

            if ( e.ctrlKey ) {

              e.preventDefault(); // prevent default behavior

              switch ( e.keyCode ) {
                case 38: // up
                  $currentTabItem.focus();
                  break;

                case 33: // pg up
                  $self.find('> ' + accTrigger).attr(setFalse);
                  $prevTab.focus().attr(setTrue);
                  break;

                case 34: // pg down
                  $self.find('> ' + accTrigger).attr(setFalse);
                  $nextTab.focus().attr(setTrue);
                  break;

                default:
                  break;
              } // end switch

            } // end if

          } // end else if

        };


        // Run setups on load
        genAcc();
        setupAcc();
        showAllSetup();


        // Events
        $self.find('> ' + accTrigger).on( 'click', panelReveal.bind(this) );
        $self.find('> ' + accPanel).on( 'click', panelClick.bind(this) );
        $self.find('> ' + accTrigger).on( 'keydown', keytrolls.bind(this) );
        $self.find('> ' + accPanel).on( 'keydown', keytrolls.bind(this) );

      }); // end: return this.each()
    }

  });  // end $.fn.extend

  // call it
  $(accWidget).a11yAccordion();

})( jQuery, this, this.document );


index.js
;(function ( w, doc, undefined ) {
	'use strict';

	/**
	 * Local object for method references
	 * and define script meta-data
	 */
	var ARIAaccordion = {};
	w.ARIAaccordion   = ARIAaccordion;

	ARIAaccordion.NS      = 'ARIAaccordion';
	ARIAaccordion.AUTHOR  = 'Scott O\'Hara';
	ARIAaccordion.VERSION = '3.2.1';
	ARIAaccordion.LICENSE = 'https://github.com/scottaohara/accessible_accordions/blob/master/LICENSE';

	var widgetClass   = 'accordion';
	var widgetTriggerClass = widgetClass + '__trigger';
	var widgetHeadingClass = widgetClass + '__heading';
	var widgetPanelClass = widgetClass + '__panel';

	var widgetHeading = '[data-aria-accordion-heading]';
	var widgetPanel   = '[data-aria-accordion-panel]'

	var idCounter = 0;

	/**
	 * Global Create
	 *
	 * This function validates that the minimum required markup
	 * is present to create the ARIA widget(s).
	 * Any additional markup elements or attributes that
	 * do not exist in the found required markup patterns
	 * will be generated via this function.
	 */
	ARIAaccordion.create = function () {
		var self;
		var panels;
		var defaultPanel = 'none';
		var headings;
		var triggers;
		var constantPanel;
		var multiPanel;
		var i;

		var widget = doc.querySelectorAll('[data-aria-accordion]');

		idCounter += 1;

		for ( i = 0; i < widget.length; i++ ) {
			self = widget[i];
			var t;

			/**
			 * Check for IDs and create arrays of necessary
			 * panels & headings for further setup functions.
			 */
			if ( !self.hasAttribute('id') ) {
				self.id = 'acc_' + idCounter + '-' + i;
			}

			/**
			 * Setup accordion classes
			 */
			self.classList.add(widgetClass);


			/**
			 * Get all panels & headings of an accordion pattern based
			 * on a specific ID > direct child selector (this will ensure
			 * that nested accordions don't get properties meant for
			 * the parent accordion, or vice-versa).
			 *
			 * If accordions are contained within an ol/ul, the selector
			 * needs to be different.
			 */
			if ( doc.querySelectorAll('#' + self.id + '> li').length ) {
				panels = doc.querySelectorAll('#' + self.id + ' li > ' + widgetPanel);
				headings = doc.querySelectorAll('#' + self.id + ' li > ' + widgetHeading);
			}
			else {
				panels = doc.querySelectorAll('#' + self.id + ' > ' + widgetPanel);
				headings = doc.querySelectorAll('#' + self.id + ' > ' + widgetHeading);
			}

			/**
			 * Check for options:
			 * data-default - is there a default opened panel?
			 * data-constant - should the accordion always have A panel open?
			 */
			if ( self.hasAttribute('data-default') ) {
				defaultPanel = self.getAttribute('data-default');
			}

			/**
			 * Accordions with a constantly open panel are not a default
			 * but if a data-constant attribute is used, then we need this
			 * to be true.
			 */
			constantPanel = self.hasAttribute('data-constant');

			/**
			 * Accordions can have multiple panels open at a time,
			 * if they have a data-multi attribute.
			 */
			multiPanel = self.hasAttribute('data-multi');

			/**
			 * If accordion panels are meant to transition, apply this inline style.
			 * This is to help mitigate a quick flash of CSS being applied to the
			 * no-js styling, and having an unwanted transition on initial page load.
			 */
			if ( self.hasAttribute('data-transition') ) {
				var thesePanels = self.querySelectorAll(widgetPanel);

				for ( t = 0; t < thesePanels.length; t++ ) {
					thesePanels[t].classList.add(widgetPanelClass + '--transition');
				}
			}

			/**
			 * Setup Panels, Headings & Buttons
			 */
			ARIAaccordion.setupPanels(self.id, panels, defaultPanel, constantPanel);
			ARIAaccordion.setupHeadingButton(headings, constantPanel);

			if ( doc.querySelectorAll('#' + self.id + '> li').length ) {
				triggers = doc.querySelectorAll('#' + self.id + ' li > ' + widgetHeading + ' .' + widgetTriggerClass);
			}
			else {
				triggers = doc.querySelectorAll('#' + self.id + ' > ' + widgetHeading + ' .' + widgetTriggerClass);
			}

			/**
			 * Now that the headings/triggers and panels are setup
			 * we can grab all the triggers and setup their functionality.
			 */
			for ( t = 0; t < triggers.length; t++ ) {
				triggers[t].addEventListener('click', ARIAaccordion.actions);
				triggers[t].addEventListener('keydown', ARIAaccordion.keytrolls);
			}
		} // for(widget.length)
	}; // ARIAaccordion.create()


	ARIAaccordion.setupPanels = function ( id, panels, defaultPanel, constantPanel ) {
		var i;
		var panel;
		var panelID;
		var setPanel;
		var constant;

		for ( i = 0; i < panels.length; i++ ) {
			panel = panels[i];
			panelID = id + '_panel_' + (i + 1);
			setPanel = defaultPanel;
			constant = constantPanel;

			panel.setAttribute('id', panelID);
			ariaHidden(panels[0], true);

			panel.classList.add(widgetPanelClass);

			/**
			 * Set the accordion to have the appropriately
			 * opened panel if a data-default value is set.
			 * If no value set, then no panels are open.
			 */
			if ( setPanel !== 'none' && parseInt(setPanel) !== NaN ) {
				// if value is 1 or less
				if ( setPanel <= 1 ) {
					ariaHidden(panels[0], false);
				}
				// if value is more than the number of panels, then open
				// the last panel by default
				else if ( (setPanel - 1) >= panels.length ) {
					ariaHidden(panels[panels.length - 1], false);
				}
				// for any other value between 2 - the last panel #, open that one
				else {
					ariaHidden(panels[(setPanel - 1)], false);
				}
			}

			/**
			 * If an accordion is meant to have a consistently open panel,
			 * and a default open panel was not set (or was not set correctly),
			 * then run one more check.
			 */
			if ( constant && setPanel === 'none' || parseInt(setPanel) === NaN ) {
				ariaHidden(panels[0], false);
			}
		}
	}; // ARIAaccordion.setupPanels


	ARIAaccordion.setupHeadingButton = function ( headings, constantPanel ) {
		var heading;
		var targetID;
		var targetState;
		var newButton;
		var buttonText;
		var i;

		for ( i = 0; i < headings.length; i++ ) {
			heading     = headings[i];
			targetID    = heading.nextElementSibling.id;
			targetState = doc.getElementById(targetID).getAttribute('aria-hidden');

			// setup new heading buttons
			newButton  = doc.createElement('button');
			buttonText = heading.textContent;
			// clear out the heading's content
			heading.innerHTML = '';
			// provide the heading with a class for styling
			heading.classList.add(widgetHeadingClass);

			newButton.setAttribute('type', 'button');
			newButton.setAttribute('aria-controls', targetID);
			newButton.setAttribute('id', targetID + '_trigger');
			newButton.classList.add(widgetTriggerClass);

			/**
			 * Check the corresponding panel to see if it was set up
			 * to be hidden or shown by default. Add an aria-expanded
			 * attribute value that is appropriate.
			 */
			if ( targetState === 'false' ) {
				ariaExpanded(newButton, true);
				isCurrent(newButton, true);

				/**
				 * Check to see if this an accordion that needs a constantly
				 * opened panel, and if the button's target is not hidden.
				 */
				if ( constantPanel ) {
					newButton.setAttribute('aria-disabled', 'true');
				}
			}
			else {
				ariaExpanded(newButton, false);
				isCurrent(newButton, false);
			}

			// Add the Button & previous heading text
			heading.appendChild(newButton);
			newButton.appendChild(doc.createTextNode(buttonText));
		}
	}; // ARIAaccordion.createButton


	ARIAaccordion.actions = function ( e ) {
		// Need to pass in if this is a multi accordion or not.
		// Also need to pass in existing trigger arrays.
		var thisAccordion = this.id.replace(/_panel.*$/g, '');
		var thisTarget = doc.getElementById(this.getAttribute('aria-controls'));
		var thisTriggers;

		if ( doc.querySelectorAll('#' + thisAccordion + '> li').length ) {
			thisTriggers = doc.querySelectorAll('#' + thisAccordion + ' li > ' + widgetHeading + ' .' + widgetTriggerClass);
		}
		else {
			thisTriggers = doc.querySelectorAll('#' + thisAccordion + ' > ' + widgetHeading + ' .' + widgetTriggerClass);
		}

		e.preventDefault();

		ARIAaccordion.togglePanel( e, thisAccordion, thisTarget, thisTriggers );
	}; // ARIAaccordion.actions()


	ARIAaccordion.togglePanel = function ( e, thisAccordion, targetPanel, triggers ) {
		var getID;
		var i;
		var thisTrigger = e.target;

		// check to see if a trigger is disabled
		if ( thisTrigger.getAttribute('aria-disabled') !== 'true' ) {

			getID = thisTrigger.getAttribute('aria-controls');

			isCurrent(thisTrigger, 'true');

			if ( thisTrigger.getAttribute('aria-expanded') === 'true' ) {
				ariaExpanded(thisTrigger, 'false');
				ariaHidden(targetPanel, 'true');
			}
			else {
				ariaExpanded(thisTrigger, 'true');
				ariaHidden(targetPanel, 'false');

				if ( doc.getElementById(thisAccordion).hasAttribute('data-constant') ) {
					ariaDisabled(thisTrigger, 'true');
				}
			}

			if ( doc.getElementById(thisAccordion).hasAttribute('data-constant') ||
					 !doc.getElementById(thisAccordion).hasAttribute('data-multi') ) {

				for ( i = 0; i < triggers.length; i++ ) {
					if ( thisTrigger !== triggers[i] ) {
						isCurrent(triggers[i], 'false');
						getID = triggers[i].getAttribute('aria-controls');
						ariaDisabled(triggers[i], 'false');
						ariaExpanded(triggers[i], 'false');
						ariaHidden(doc.getElementById(getID), 'true');
					}
				}
			}
		}
	};


	ARIAaccordion.keytrolls = function ( e ) {
		if ( e.target.classList.contains(widgetTriggerClass) ) {
			var keyCode = e.keyCode || e.which;

			// var keyUp = 38;
			// var keyDown = 40;
			var keyHome = 36;
			var keyEnd = 35;

			var thisAccordion = this.id.replace(/_panel.*$/g, '');
			var thisTriggers;

			if ( doc.querySelectorAll('#' + thisAccordion + '> li').length ) {
				thisTriggers = doc.querySelectorAll('#' + thisAccordion + ' li > ' + widgetHeading + ' .' + widgetTriggerClass);
			}
			else {
				thisTriggers = doc.querySelectorAll('#' + thisAccordion + ' > ' + widgetHeading + ' .' + widgetTriggerClass);
			}


			switch ( keyCode ) {
				/**
				 * keyUp & keyDown are optional controls
				 * for accordion components.
				 */
				// case keyUp:
				// 	if ( doc.getElementById(thisAccordion).hasAttribute('data-up-down') ) {
				// 		e.preventDefault();
				// 		// optional up arrow controls
				// 	}
				// 	break;

				// case keyDown:
				// 	if ( doc.getElementById(thisAccordion).hasAttribute('data-up-down') ) {
				// 		e.preventDefault();
				// 		// optional down arrow control
				// 	}
				// 	break;

				/**
				 * keyEnd/Home are optional functions that may not be inherently known
				 * to most users and, in the case of END, conflict with expected
				 * usage of that key with NVDA.
				 */
				case keyEnd:
					e.preventDefault();
					thisTriggers[thisTriggers.length - 1].focus();
					break;

				case keyHome:
					e.preventDefault();
					thisTriggers[0].focus();
					break;

				default:
					break;
			}
		}
	}; // ARIAaccordion.keytrolls()


	/**
	 * Initialize Accordion Functions
	 * if expanding this script, place any other
	 * initialize functions within here.
	 */
	ARIAaccordion.init = function () {
		ARIAaccordion.create();
	};


	/**
	 * Helper Functions
	 * Just to cut down on the verboseness of some declarations
	 */
	var ariaHidden = function ( el, state ) {
		el.setAttribute('aria-hidden', state);
	};

	var ariaExpanded = function ( el, state ) {
		el.setAttribute('aria-expanded', state);
	};

	var ariaDisabled = function ( el, state ) {
		el.setAttribute('aria-disabled', state);
	};

	var isCurrent = function ( el, state ) {
		el.setAttribute('data-current', state);
	};

	// go go JavaScript
	ARIAaccordion.init();

})( window, document );
